Speed up safe_strtod and safe_strtof functions by using double-conversion library

akindyakov opened the pull request:
#11713

akindyakov committed:
Add double-conversion library to third_party

Link: https://github.com/google/double-conversion
Version: v3.0.0
With patch (add case insensivity for special values): https://github.com/google/double-conversion/pull/47

akindyakov committed:
Strnlen function to str_util.h
Returns the length of the given null-terminated byte string 'str'.
Returns strsz if the null character was not found in the first
'strsz' bytes of 'str'.

akindyakov committed:
Use double-conversion lib for safe_strto{d,f} implementation

tensorflow-jenkins commented:
Can one of the admins verify this patch?

AKindyakov commented:
@rmlarsen, performance improvement is so attractive :)
There is benchmark: https://gist.github.com/AKindyakov/2c528beccb11a24c1e6ec02ec12df47e
DC is significantly faster.

rmlarsen commented:
@AKindyakov thanks. That is indeed a nice speedup. 
@martinwicke @vrv what are your opinions? Should we introduce this new dependency or try to speed up decode_csv in another way?

rmlarsen commented:
We will investigate and see what direction to take.

drpngx commented:
@rmlarsen any decision on this?

I think we might be open sourcing the google library, which is probably faster than the TF re-implementation.

rmlarsen commented:
@drpngx I have not had time to pursue this.

AKindyakov commented:
JFYI patch about case insensitivity for special values was accepted and merged into the master branch.
https://github.com/google/double-conversion/pull/47

drpngx commented:
Oh, I hadn't realized this was from google. @martinwicke wdyt about taking on this additional dependency?

martinwicke commented:
I have no problem with it. Seems lightweight enough. We should double-check whether it supports all platforms, but I see MSVC files there, so I presume the answer is yes.

drpngx commented on third_party/double-conversion/add_case_insensitive_nan.patch:
```diff
@@ -0,0 +1,188 @@
+--- a/double-conversion/double-conversion.cc
++++ b/double-conversion/double-conversion.cc
+@@ -26,6 +26,7 @@
+ // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
Please undo, as per comment, case insensitivity should already be there.

drpngx commented on tensorflow/core/lib/strings/str_util.cc:
```diff
@@ -384,5 +384,13 @@ bool SplitAndParseAsFloats(StringPiece text, char delim,
                                     result);
 }
 
+size_t Strnlen(const char* str, size_t strsz) {
```
`const size_t string_max_len` instead of `strsz`

sb2nov commented:
@AKindyakov can you look at the merge conflict and the PR comments

AKindyakov commented:
Yes, sure, I'll do it this night.

rmlarsen commented:
I honestly don't have time to look at this right now, unassigning myself. @sb2nov could you re-assign, please?

akindyakov committed:
Path for double-conversion library is no longer needed, just use new version of this lib

akindyakov committed:
Use more obvious arguments names for Strnlen function

sb2nov commented:
@AKindyakov there is a merge conflict. Can you take a look.

akindyakov committed:
Merge master from upstream

AKindyakov commented:
@sb2nov, it's done

sb2nov commented:
Jenkins, test this please.

akindyakov committed:
Fixed DoubleToBuffer function in case of [full_precision_needed]
use safe_strtod insetead of raw inplementation

AKindyakov commented:
Jenkins, test this, please.

akindyakov committed:
add double-conversion compilation in makefile builds
- created compilation script: compile_double_conversion.sh
- added running compile_double_conversion.sh in {build_all_android.sh, build_all_ios.sh, build_all_linux.sh}
- added linking libdouble-conversion.a to build in tensorflow/contrib/makefile/Makefile

AKindyakov commented:
@sb2nov, I've fixed build and test problems. May I ask you to run full tests set for it?

drpngx commented:
Jenkins, test this please.

On Oct 4, 2017 2:21 AM, "Alexander" <notifications@github.com> wrote:

> @sb2nov <https://github.com/sb2nov>, I've fixed build and test problems.
> May I ask you to run full tests set for it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/12102#issuecomment-334097757>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AT_SbfXFxLiqu791N0g8cUS1PTNIHweHks5so04RgaJpZM4Owazw>
> .
>

akindyakov committed:
Build double-conversion directly inside tensorflow/contrib/makefile/Makefile

akindyakov committed:
Put double_conversion to the rest of targets including number.* as src

AKindyakov commented:
I've fixed makefile build and have fixed up dependencies lists in ``tensorflow/core``.
May I ask you to run full tests on it?

akindyakov committed:
Sort external dependencies and make Sanity Checks happy

AKindyakov commented:
Fixed problems with sanity check

AKindyakov commented:
Can I ask someone to run the full test set for it?

cwhipkey commented:
jenkins, test this please

cwhipkey commented on tensorflow/core/BUILD:
```diff
@@ -1092,6 +1096,7 @@ cc_library(
     deps = [
         ":protos_all_cc_impl",
         "//third_party/eigen3",
+        "@double_conversion//:double-conversion",
```
It's a little unfortunate that mobile builds will end up using it, when I assume they don't need the extra speed of parsing (since they will only be parsing a few attribute values, not large files of text).  Do you know how large the double-conversion library is when compiled for android?

(one option if the net change in binary size would be too large is to have the double conversion library only used by the op that needs it for speed).

AKindyakov replied:
It's not so large actually. The size of the compiled static library is 486KiB on my laptop.

AKindyakov replied:
Stripped binary build with optimisation level 3 has size 77KiB. By the way, I can figure out what can we throw out from the library in the android build.

cwhipkey commented on tensorflow/core/lib/strings/numbers.cc:
```diff
@@ -165,8 +112,8 @@ char* DoubleToBuffer(double value, char* buffer) {
 
-    full_precision_needed =
-        locale_independent_strtonum<double>(buffer, nullptr) != value;
+    auto parsed_value = double{};
```
why this change?  (also, we could possibly use DoubleToStringConverter to implement all of DoubleToBuffer?)  (but see the question about android code size as well).

AKindyakov replied:
The previous implementation of DoubleToBuffer function was using old locale_independent_strtonum<double> which is replaced by double_conversion::StringToDoubleConverter with slightly differen interface.

But it is a good point to use only one function to convert str to double everywhere here. Consider double_conversion::StringToDoubleConverter only as hidden implementation details of safe_strto{d,f}.

This is a good idea to use double-conversion for {Double,Float}ToBuffer functions. I can do it here or in another pull request.

cwhipkey commented on tensorflow/core/lib/strings/numbers_test.cc:
```diff
@@ -278,6 +279,30 @@ TEST(safe_strtof, Float) {
   EXPECT_EQ(-42.0f, result);
 
```
maybe test case of leading and trailing whitespace as well

AKindyakov replied:
Sure, I'll do it. Thank you.

cwhipkey commented on tensorflow/core/lib/strings/numbers.cc:
```diff
@@ -32,72 +35,16 @@ namespace tensorflow {
+    const static double_conversion::StringToDoubleConverter converter(
+        double_conversion::StringToDoubleConverter::ALLOW_LEADING_SPACES
+        | double_conversion::StringToDoubleConverter::ALLOW_HEX
+        | double_conversion::StringToDoubleConverter::ALLOW_OCTALS
```
do we need to support OCTALS?

AKindyakov replied:
I don't know. Should I remove it for the sake of nice little interface?

cwhipkey replied:
sgtm, since we didn't support before I'd rather not add support if we don't really need it.

AKindyakov replied:
ok, I'll remove it.

akindyakov committed:
Add test cases with trailing and leading whitespace characters to //tensorflow/core:lib_strings_numbers_test

akindyakov committed:
Remove octal numbers support from safe_strtod and safe_strtof

akindyakov committed:
Add double-conversion library to the cmake build

AKindyakov commented:
@cwhipkey I fixed up ```CMake``` build, remove octal number support and add test cases with trailing and leading characters. Could you please run tests on it?

I was looking for a code in double conversion library which we can throw out in the android build. But this library is very solid and there is nothing to cut off.

cwhipkey commented:
+ @petewarden  - pete, this will increase binary size some on  mobile (maybe not the full 77k, but it will probably be an increase).  What do you think we should do?   Some options are:

1. use ifdef to use the old code on mobile.  This clearly would cause no binary increase on mobile, but makes another code path that will be less tested and it's a bit unfortunate to have 2 code paths.

2. similar to 1., but do it in a way that can be tested off android.

3. add in the performant library and code only use by the kernels, but not in the string library.  This seems not great, since you would need to call the fast library from the kernels, instead of just using the string library.

4. test more fully and maybe it won't be a big change, or we can find savings elsewhere.

samtaylor159 commented:
Is there any update on this? This issue seems to be a current bottleneck in my code, it adds a huge overhead when reading my data.

cwhipkey commented:
There is a pull request for it:
https://github.com/tensorflow/tensorflow/pull/12102

petewarden reviewed (approved):
Sorry for the delay! We may end up figuring out how to avoid the extra dependency on mobile as an option after this goes in, but I don't want that to block this change, so approving from my side.

martinwicke committed:
Merge branch 'master' into speed_up_csv_reader

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
There are still problems with the Windows build, relevant error:

```
17:22:02 LINK : fatal error LNK1181: cannot open input file 'double_conversion\src\double_conversion\double-conversion\double-conversion.lib' [C:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\cmake_build\proto_text.vcxproj]
```

martinwicke commented:
@mrry, do you know how to fix this?

mrry commented:
At a guess, maybe the static library has a different filename on Windows? Maybe there's a missing `lib/` in the definition of `double_conversion_STATIC_LIBRARIES`?

AKindyakov commented:
Is it me suppose to fix this?
I don't have any WIN machine for the test.

drpngx committed:
Prepend lib/ to cmake lib path as per mrry instructions

See https://www.github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cmake/external/zlib.cmake#L24

drpngx commented:
@AKindyakov yes usually the committer ensures that everything builds, with our help. For this one, I pushed a commit that I think should fix it.

drpngx commented:
@AKindyakov yes usually the committer ensures that everything builds, with our help. For this one, I pushed a commit that I think should fix it.

drpngx commented:
@AKindyakov yes usually the committer ensures that everything builds, with our help. For this one, I pushed a commit that I think should fix it.

AKindyakov commented:
@drpngx thanks a lot!

drpngx commented:
Looks like there's a sycl breakage at `master`. Let's wait.

martinwicke committed:
Merge branch 'master' into speed_up_csv_reader

martinwicke commented:
Jenkins, test this please.

AKindyakov commented:
@martinwicke, @drpngx have you any idea what is wrong with the build. There are no errors in build log only warnings which are not related to my changes at all.

drpngx commented:
The mac test failed with a timeout.

The CMake failed with the same error as before: 
```
22:31:05 LINK : fatal error LNK1181: cannot open input file 'double_conversion\src\double_conversion\double-conversion\lib\double-conversion.lib' [C:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\cmake_build\proto_text.vcxproj]
```
The other errors don't appear to be related. Let's try again.

akindyakov committed:
Add  to double-conversion win32 build output path: fix up windows CMake build

AKindyakov commented:
@martinwicke, @drpngx, I believe I fixed up windows build. May I ask you to test it, please? (sorry for the delay - busy time)

martinwicke commented:
Jenkins. Test this please.

sb2nov commented:
@martinwicke is there a way to fix the CLA bot here. It was approved yes but the check seems to be pending.

martinwicke commented:
It looks to me like it is stuck. I will override. I guess the XLA tests are known to be broken?

@willnorris FYI, not sure how to get it out of its funk.

willnorris commented:
There's no obvious reason that I can see that would have caused this to get stuck, but I've manually approved it and will look into the cause.

AKindyakov commented:
Thanks!

caisq commented:
Just as a heads-up: This contribution is causing difficulties when the code is sync'ed to Google internal. We are considering reverting it first to unblock the code sync'ing, and maybe finding a way to make a similar change in internal.

cc @gunan
